### PR TITLE
ECT: CheckAreaLedger

### PR DESCRIPTION
**Pull request by aarontran**
When using embedded boundary with enlarged-cell technique (ECT) solver, check that global face area is conserved (lent area = borrowed area) to within an automatically-determined round-off error tolerance. Also check that S_mod > 0 for all participating faces to guard against an edge case wherein multiple unstable cells can "overdraft" a neighbor cell's available area.

This PR was created to help test PR #7104, which fixes area-conservation bugs related to cross-box synchronization of lent/borrowed face area.

During review of PR #7104, it was also noticed that S_mod < 0 on a face may occur due to at least one pathological case of an under-resolved embedded boundary with cut-cells at AMReX box corners.  The present PR simply aborts and does not attempt to re-work the ECT algorithm for a relatively obscure edge case.

Questions and notes for reviewers:
* Log with `verbose >=1` or `> 1` ?
* @ax3l 's comment on issue #2257 suggests that ECT boundary face extensions may be recomputed dynamically as a simulation evolves, either PEC boundary moving or AMReX regridding.  Do you think this is a serious practical concern?  If so, this PR may need to be edited to prevent check from running too often: maybe 1x at init, at some regular interval, only if user requests auditing, etc...
* Merge this PR after #7104 (CI compiles will break otherwise)

Example logging if checks passed for a 2D simulation:

```
                                   ___
    __        __             ___  /  /
    \ \      / /_ _ _ __ _ __\  \/  /
     \ \ /\ / / _` | '__| '_ \\    /
      \ V  V / (_| | |  | |_) /    \
       \_/\_/ \__,_|_|  | .__/__/\  \
                        |_|       \__\

Level 0: dt = 1.179327168e-10 ; dx = 0.05 ; dz = 0.05
--- INFO    : Embedded Boundary: ECT CheckAreaLedger after 1-way pass (idim=1)
             imbalance sum(S-S_mod) = 0.000000e+00 tolerance = 5.107026e-15 OK.
             Modified faces min(S_mod) = 1.000000e-03 > 0 OK.
--- INFO    : Embedded Boundary: ECT CheckAreaLedger after 8-way pass (idim=1)
             imbalance sum(S-S_mod) = -2.249720e-18 tolerance = 5.129230e-15
             OK. Modified faces min(S_mod) = 2.500000e-04 > 0 OK.

Grids Summary:
  Level 0   16 grids  1024 cells  100 % of domain
            smallest grid: 8 x 8  biggest grid: 8 x 8
```

Example logging if a check fails:

```
0::Assertion `balance && no_overdraft' failed, file "[...]/warpx/Source/EmbeddedBoundary/WarpXFaceExtensions.cpp", line 203, Msg:
### ERROR   : Embedded Boundary: ECT CheckAreaLedger after 1-way pass (idim=1)
#            imbalance sum(S-S_mod) = -1.734723e-18 tolerance = 9.436896e-16
#            OK. Modified faces min(S_mod) = -5.000000e-04 but expected > 0 for
#            all faces, exiting! Try resolving embedded boundary with more
#            cells, varying AMReX grid layout, and/or filing bug report.
```

### CODE REVIEW

File: Source/EmbeddedBoundary/WarpXFaceExtensions.cpp [hunk 1]
```diff
@@ -566,6 +708,8 @@ WarpX::ComputeFaceExtensions ()
     ::init_borrowing(m_borrowing[maxLevel()], Bfield);
     ComputeOneWayExtensions();
 
+    ::CheckAreaLedger("after 1-way pass", m_fields, m_ect_face_owner_mask, maxLevel(), verbose);
```
Comment (clarkse-he) [inline Source/EmbeddedBoundary/WarpXFaceExtensions.cpp:713]: On #7104 this needs to sit after `sync_lent_areas()` (the call that follows `ComputeOneWayExtensions(lent_area)` there), not directly after the pass: before the reduction the owners' `S_mod` lacks the remote debits and `net` comes out as minus the cross-box lent area.

**Review by clarkse-he — COMMENTED** [review]
Thanks Aaron — this is a good invariant to pin down, and it has already paid for itself: it is what caught the `boxArray(maxLevel())` init-order bug in #7104. The reductions are exact where they need to be (owner-only sum over faces with `S != S_mod`, global min of `S_mod`), the tolerance scales with the number of modified faces so large 3D runs will not false-trip, and the assert fires after collective reductions so every rank sees the same verdict.

Two sequencing points with #7104 before this can go in:

1. **Placement on the #7104 tree.** Here the calls follow `ComputeOneWayExtensions()` / `ComputeEightWaysExtensions()` directly. On #7104 each pass is followed by `sync_lent_areas()`, which applies the remote lent-area debits to the owners' `area_mod` and `OverrideSync`s the copies. The ledger only balances *after* that reduction — before it the owners are short by exactly the area remote fabs borrowed from them, and the check would abort on any multi-box layout. A `merge-tree` of this branch onto #7104's head reports a content conflict in this file, so a rebase is needed anyway; when you do it, please put the two `::CheckAreaLedger(...)` calls after the two `sync_lent_areas()` calls (which is where your three seam tests must have had them). Marking this "blocked by #7104" until then keeps CI honest.
2. **Regrid / recompute (your question in the description):** yes, `ComputeFaceExtensions` is re-run on every load-balance regrid (`WarpXRegrid.cpp` → `InitializeEBGridData` → `ComputeFaceExtensions`), so the check runs then too. Cost is three scalar reductions per direction per pass, so I would not gate it on anything. Moving PEC / dynamic EB is not something the ECT path supports today, so nothing to worry about there.

`verbose >= 1` is fine by me — a few lines at init and at each regrid, the same cadence as the "Faces to be extended" messages this function already emits. Inline nits are optional. I'll approve once it is rebased onto #7104.

File: Source/EmbeddedBoundary/WarpXFaceExtensions.cpp [hunk 2]
```diff
@@ -578,6 +722,9 @@ WarpX::ComputeFaceExtensions ()
     );
 
     ComputeEightWaysExtensions();
+
+    ::CheckAreaLedger("after 8-way pass", m_fields, m_ect_face_owner_mask, maxLevel(), verbose);
```
Comment (clarkse-he) [inline Source/EmbeddedBoundary/WarpXFaceExtensions.cpp:728]: Same here: after the second `sync_lent_areas()` on #7104 (before `shrink_borrowing`, which does not touch `S_mod`, and before the BCK correction, which does write `face_areas`).

File: Source/EmbeddedBoundary/WarpXFaceExtensions.cpp [hunk 3]
```diff
@@ -69,7 +76,142 @@ namespace
         return sums;
     }
 
+
+    /**
+    * \brief Check that face extensions conserved area and did not overdraft
+    * any lending face to S_mod <= 0.  The global sum of original areas S must
+    * equal sum of modified areas S_mod to within a round-off error tolerance.
+    * This routine must be called before BCK correction, which overwrites
```
Comment (clarkse-he) [inline Source/EmbeddedBoundary/WarpXFaceExtensions.cpp]: Verified: `ApplyBCKCorrection` writes `S(i,j,k) = ComputeSStab(...)` into `face_areas`, so this ordering note is exactly right. Suggest appending "…and after any cross-box reduction of `area_mod` (see `sync_lent_areas` in `ComputeFaceExtensions` once #7104 is in)".

File: Source/EmbeddedBoundary/WarpXFaceExtensions.cpp [hunk 4]
```diff
@@ -69,7 +76,142 @@ namespace
...
+                            const amrex::Real d = S(i,j,k) - S_mod(i,j,k);
+                            if (owner(i,j,k) == 0 || d == amrex::Real(0.)) {
+                                // skip non-owned faces on shared nodal planes
+                                // skip faces that neither lend nor borrow
+                                return { amrex::Real(0.),
+                                         amrex::Real(0.),
+                                         std::numeric_limits<amrex::Real>::max() };
+                            } else {
+                                return { d,
+                                         std::max(S(i,j,k),S_mod(i,j,k)),
```
Comment (clarkse-he) [inline Source/EmbeddedBoundary/WarpXFaceExtensions.cpp:148]: Nit: `amrex::max` rather than `std::max` in device code (WarpX convention; `std::max` only works under NVCC because AMReX passes `--expt-relaxed-constexpr`). The `numeric_limits` constant three lines up is fine as a constexpr, but hoisting it to a `constexpr auto huge = std::numeric_limits<amrex::Real>::max();` above the lambda reads better and lets you drop the `<algorithm>` include.

File: Source/EmbeddedBoundary/WarpXFaceExtensions.cpp [hunk 5]
```diff
@@ -69,7 +76,142 @@ namespace
...
+                auto r = reduce_data.value();
+                net       = amrex::get<0>(r);
+                S_max_sum = amrex::get<1>(r);
+                S_mod_min = amrex::get<2>(r);
+                amrex::ParallelDescriptor::ReduceRealSum(net);
+                amrex::ParallelDescriptor::ReduceRealSum(S_max_sum);
+                amrex::ParallelDescriptor::ReduceRealMin(S_mod_min);
+            }
+
+            // Round-off error may arise from the parallel reduction or from
```
Comment (clarkse-he) [inline Source/EmbeddedBoundary/WarpXFaceExtensions.cpp:164]: The bound is sound, and the important property is that it scales with the number of modified faces through `S_max_sum`, so big 3D runs cannot false-trip. Worth stating that explicitly, plus why the reduction's own round-off stays small: lenders and borrowers are spatially adjacent, so partial sums cancel locally. The `1e2` prefactor then covers the debit arithmetic (a handful of ops per face).

File: Source/EmbeddedBoundary/WarpXFaceExtensions.cpp [hunk 6]
```diff
@@ -69,7 +76,142 @@ namespace
...
+                    auto       const &S     = S_mf    ->const_array(mfi);
+                    auto       const &S_mod = S_mod_mf->const_array(mfi);
+                    auto       const &owner = owner_mf->const_array(mfi);
+
+                    reduce_ops.eval(box, reduce_data,
+                        [=] AMREX_GPU_DEVICE(int i, int j, int k)
+                        -> amrex::GpuTuple<amrex::Real, amrex::Real, amrex::Real> {
+
+                            const amrex::Real d = S(i,j,k) - S_mod(i,j,k);
+                            if (owner(i,j,k) == 0 || d == amrex::Real(0.)) {
```
Comment (clarkse-he) [inline Source/EmbeddedBoundary/WarpXFaceExtensions.cpp:140]: Exact-zero compare is intended here (a face that neither lent nor borrowed is bit-identical to `S`), worth one word in the comment so nobody later "fixes" it with a tolerance. In single precision a lender debited by less than half an ulp of `S` reads `d == 0` while the borrower recorded the credit; that lands in `net` at the eps level, inside the tolerance, so no issue.

File: Source/EmbeddedBoundary/WarpXFaceExtensions.cpp [hunk 7]
```diff
@@ -69,7 +76,142 @@ namespace
...
+                const bool balance = amrex::Math::abs(net) <= rtol*S_max_sum;
+                if (balance) {
+                    msg << " OK.";
+                } else {
+                    msg << " exceeded, borrowed and lent area do not balance, exiting!";
+                }
+
+                msg << " Modified faces min(S_mod) = " << S_mod_min;
+
+                const bool no_overdraft = S_mod_min > amrex::Real(0.);
```
Comment (clarkse-he) [inline Source/EmbeddedBoundary/WarpXFaceExtensions.cpp:193]: Optional follow-up, not for this PR: `> 0` catches the catastrophic overdraft. The lending rule is stronger — a lender has to stay stable, `S_mod >= S_stab(face)` — and checking that would also catch a lender pushed below stability but still positive, which later shows up as a spuriously large `1/S_mod` in `EvolveBCartesianECT`. `ComputeSStab<idim>` already provides the threshold.

File: Source/EmbeddedBoundary/WarpXFaceExtensions.cpp [hunk 8]
```diff
@@ -69,7 +76,142 @@ namespace
...
+                }
+
+                if (!balance || !no_overdraft ) {
+                    msg << " Try resolving embedded boundary with more cells, varying AMReX grid layout, and/or filing bug report.";
+                }
+
+                WARPX_ALWAYS_ASSERT_WITH_MESSAGE(balance && no_overdraft,
+                                                 msg.str());
+
+                if (verbose >= 1) {
```
Comment (clarkse-he) [inline Source/EmbeddedBoundary/WarpXFaceExtensions.cpp:207]: `>= 1` is fine (see summary). If it turns out noisy on runs with frequent load-balance regrids we can move it to `> 1` later.

File: Source/EmbeddedBoundary/WarpXFaceExtensions.cpp [hunk 9]
```diff
@@ -69,7 +76,142 @@ namespace
...
+            amrex::Real S_mod_min;  // min(S_mod) verify no overdraft
+            {
+                amrex::ReduceOps< amrex::ReduceOpSum,
+                                  amrex::ReduceOpSum,
+                                  amrex::ReduceOpMin > reduce_ops;
+                amrex::ReduceData< amrex::Real,
+                                   amrex::Real,
+                                   amrex::Real > reduce_data(reduce_ops);
+
+                for (amrex::MFIter mfi(*S_mf); mfi.isValid(); ++mfi) {
```
Comment (clarkse-he) [inline Source/EmbeddedBoundary/WarpXFaceExtensions.cpp:129]: For a reduction this is fine: no OpenMP region, and the owner mask removes the shared-plane double count that `validbox()` on a face-centred fab would otherwise give. If you want it to look like its neighbours, `amrex::TilingIfNotGPU()` + `tilebox()` is what `ApplyBCKCorrection` uses.

File: Source/EmbeddedBoundary/WarpXFaceExtensions.cpp [hunk 10]
```diff
@@ -16,10 +16,17 @@
 
 #include <AMReX_Functional.H>
 #include <AMReX_GpuAtomic.H>
+#include <AMReX_Math.H>
 #include <AMReX_Scan.H>
 #include <AMReX_iMultiFab.H>
 #include <AMReX_MultiFab.H>
 
+#include <algorithm>
```
Comment (clarkse-he) [inline Source/EmbeddedBoundary/WarpXFaceExtensions.cpp:24]: Only needed for `std::max`; can go if you switch to `amrex::max` below.

File: Source/EmbeddedBoundary/WarpXFaceExtensions.cpp (hunk 7 above)
Comment (aarontran) [reply Source/EmbeddedBoundary/WarpXFaceExtensions.cpp:193]: I don't see evidence that `S_mod >= S_stab` is required for lending faces?
* `ComputeOneWayExtensions` and `ComputeEightWaysExtensions` both allow taking from neighbor faces so long as final `S_mod > 0`, and this logic was in place prior to your recent series of pull requests.
* Xiao+ (2004 IEEE page 552) says: "... the enlargement in Fig. 1(b) is sufficient only if $S_{xy}(i+1, j, k) + S_{xy}(i, j-1, k) > S^{ext}$..." where $S^{ext}$ is the area to be debited, and $S_{xy}(...)$ is the area of the neighboring faces in the given example.

File: Source/EmbeddedBoundary/WarpXFaceExtensions.cpp (hunk 4 above)
Comment (aarontran) [reply Source/EmbeddedBoundary/WarpXFaceExtensions.cpp:148]: Accepted `huge` hoist in 7db40da1a ; declined `max` namespace change because
* AMReX source documentation in `Src/Base/AMReX_Algorithm.H` says "This function was added to AMReX to support GPU before `std::max` was constexpr in C++14. `std::max` can now be used directly instead."
* grep on warpx `Source/` folder suggests that `std::max` is used ~3x more than `amrex::max`

File: Source/EmbeddedBoundary/WarpXFaceExtensions.cpp (hunk 3 above)
Comment (aarontran) [reply Source/EmbeddedBoundary/WarpXFaceExtensions.cpp]: Accepted in 7db40da1a

File: Source/EmbeddedBoundary/WarpXFaceExtensions.cpp (hunk 5 above)
Comment (aarontran) [reply Source/EmbeddedBoundary/WarpXFaceExtensions.cpp:164]: Declined (desperately fighting Claude's verbosity in all codebases)

File: Source/EmbeddedBoundary/WarpXFaceExtensions.cpp (hunk 6 above)
Comment (aarontran) [reply Source/EmbeddedBoundary/WarpXFaceExtensions.cpp:140]: Accepted in 7db40da1a

File: Source/EmbeddedBoundary/WarpXFaceExtensions.cpp (hunk 9 above)
Comment (aarontran) [reply Source/EmbeddedBoundary/WarpXFaceExtensions.cpp:129]: Tentative decline for simplicity, keep loop structure matched to 1- and 8-way passes

File: Source/EmbeddedBoundary/WarpXFaceExtensions.cpp (hunk 10 above)
Comment (aarontran) [reply Source/EmbeddedBoundary/WarpXFaceExtensions.cpp:24]: Declined

**Comment by aarontran** [comment]
Thanks for review, accepted some suggestions in commit 7db40da1a & replies in-line above.

File: Source/EmbeddedBoundary/WarpXFaceExtensions.cpp (hunk 7 above)
Comment (clarkse-he) [reply Source/EmbeddedBoundary/WarpXFaceExtensions.cpp:193]: You're right, and I withdraw that one. I had the lender's stability wrong: in `EvolveBCartesianECT` the intruded face is updated as `B -= dt*Venl/S` over its full original area, with `Rho*S_mod` entering only as the weight of its own contribution to `Venl`, so its stability is that of the original face and does not depend on `S_mod`. The passes' rule (test-and-subtract on `S_mod - S_ext > 0`, and the `S_red - patch <= 0` rejection in the eight-way split) is exactly `S_mod > 0`, which is what the check enforces. Thanks for the Xiao et al. pointer.

**Comment by clarkse-he** [comment]
Cleanup looks good and the declines are fine by me; the only thing left from my side is the rebase onto #7104 with the two calls after `sync_lent_areas()`, then I'll approve.

File: Source/EmbeddedBoundary/WarpXFaceExtensions.cpp [hunk 11]
```diff
@@ -69,7 +76,144 @@ namespace
...
+        // In 2D we only need the case idim=1
+        for (int idim = 1; idim < AMREX_SPACEDIM; ++idim) {
+#elif defined(WARPX_DIM_3D)
+        for (int idim = 0; idim < AMREX_SPACEDIM; ++idim) {
+#else
+            WARPX_ABORT_WITH_MESSAGE(
+                "CheckAreaLedger: Only implemented in 2D3V and 3D3V");
 #endif
+            auto* const S_mf     = all_fields.get(FieldType::face_areas, Direction{idim}, max_level);
+            auto* const S_mod_mf = all_fields.get(FieldType::area_mod,   Direction{idim}, max_level);
```
Comment (clarkse-he) [inline Source/EmbeddedBoundary/WarpXFaceExtensions.cpp:114]: clang-tidy `readability-qualified-auto`: `all_fields` is const, so `get()` returns a pointer to const.

```suggestion
            const auto* const S_mf     = all_fields.get(FieldType::face_areas, Direction{idim}, max_level);
            const auto* const S_mod_mf = all_fields.get(FieldType::area_mod,   Direction{idim}, max_level);
```

**Review by clarkse-he — COMMENTED** [review]
Heads-up from carrying this under #7105: clang-tidy (the 2D job; 3D/RZ get cancelled by fail-fast) flags these two declarations with `readability-qualified-auto`, since `all_fields` is a `const MultiFabRegister&` and its `get()` returns a pointer to const. Your CI has not reached clang-tidy yet because of the `m_ect_face_owner_mask` compile error on the unrebased head, so it will show up after the rebase onto #7104.
